### DISCUSSION TITLE
Support of YOLOX

### DISCUSSION THREAD

**Discussion by ZwwWayne** [open-mmlab/mmdetection#5659] [Ideas] [answered]
Dear MMDet contributors,

First, on behalf of the MMDet team, thanks for the support along these years.

We find that the community has enthusiasm about supporting YOLOX of MMDetection. And there is already a PR to support the inference in https://github.com/open-mmlab/mmdetection/pull/5656. Considering that, we plan to open a discussion about YOLOX here, including any discussion in the implementation process, the decomposition of the modules, and the role of each contributor.

Thanks again,
Wenwei
On behalf of the MMDet team

**Comment by shinya7y** [comment]
At first glance, the original YOLOX implementation differs from mmdet's style (e.g., `YOLOX/yolox/exp/`, the neck includes the backbone), but the code can be restructured for mmdet as in https://github.com/open-mmlab/mmdetection/pull/5656/commits/370019e0b9a86c7b32539d9000279f5c2a22ae66

To my current understanding, tasks can be divided as below:

- Support inference
  - Add models
  - Restructure code for mmdet
  - Debug
  - Refactor backbone and neck
  - Refactor head and postprocess
  - Add unit tests
  - Add docstring
- Support training
  - Add mosaic/mixup augmentation
  - Add other augmentation
  - Add augmentation scheduler
  - Add unit tests
  - Train models
  - Debug
  - Add docstring
- Support deployment

In my opinion, it is better to release inference code first because it enables the community to use the backbone and neck; and to benchmark inference speed (YOLOX vs. other detectors, MMDet repo vs. YOLOX repo) and improve speed.
Mosaic/mixup augmentation also could be released before reproducing the full YOLOX.

Please note that it will be important to benchmark and keep inference speed in refactoring.

**Reply by ZwwWayne** [reply]
Hi @shinya7y ,
Thanks for your suggestion. That is basically the steps we are going to do.  But we do not merge those PRs into the master because we want to make the code merged into master branch more stable and simplified. Finally, since we have aligned the training accuracy. We are going to merge the PR in to master. The plan is as the following.

First, all PRs will be merged into `yolox_final` branch. The total feature are decomposed into 10+ action items, including the support of model, data transform, and training hooks. We will also support the model inference first in the `yolox_final` branch.

After these PRs are merged into `yolox_final`, we will take 2-3 days to benchmark the training and finally merge the `yolox_final` branch into master **without squashing the commits**. The project progress can be seen in https://github.com/open-mmlab/mmdetection/projects/2.

As your suggestion is valuable to us, would you like to participate in the review process in those PRs? Meanwhile, if you have time and are interested, we would also like to invite you to support the ONNX export of YOLOX.

**Reply by shinya7y** [reply]
I express my respect for the contributors' efforts toward the incredibly fast alignment of training accuracy.

My response may be slow depending on other tasks. If you don't mind it, please send me review requests.
I'm not so interested in the ONNX export so far.

**Comment by RangiLyu** [comment, marked as answer]
# Update

We have reimplemented YOLOX based on #5705.

Training result of YOLOX tiny is:
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.319
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.492
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.341
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.125
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.353
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.486
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.452
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.452
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.452
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.191
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.512
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.657

```

The next step is to divide the huge PR into different parts of modules. The divide of work has been updated in https://github.com/open-mmlab/mmdetection/projects/2
